reframe/utility/versioning.py

```diff
@@ -0,0 +1,56 @@
+import functools
+
+
+@functools.total_ordering
+class Version:
+    def __init__(self, version):
+        base_part, *dev_part = version.split('-dev')
+        try:
+            major, minor, *patch_part = base_part.split('.')
+        except ValueError:
+            raise ValueError('invalid version string: %s' % version) from None
+
+        patch_level = patch_part[0] if patch_part else 0
+
+        try:
+            self._major = int(major)
+            self._minor = int(minor)
+            self._patch_level = int(patch_level)
+            self._dev_number = int(dev_part[0]) if dev_part else None
+        except ValueError:
+            raise ValueError('invalid version string: %s' % version) from None
+
+    def _value(self):
+        return 1000*self._major + 100*self._minor + self._patch_level
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+
+        return (self._value() == other._value() and
+                self._dev_number == other._dev_number)
+
+    def __gt__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+
+        if self._value() != other._value():
+            return self._value() > other._value()
+        elif self._dev_number is None and other._dev_number is None:
+            return self._value() > other._value()
+        elif self._dev_number is not None and other._dev_number is None:
+            return False
+        elif self._dev_number is None and other._dev_number is not None:
+            return True
+        elif self._dev_number is not None and other._dev_number is not None:
+            return self._dev_number > other._dev_number
```

Comment:
To avoid all these elifs, I would use the trick of converting the dev number to infinity, which you have used originally, but without using `math` module:

```python
if self._value() > other._value():
    return True

if self._value() < other._value():
    return False

# Both versions are equal here; compare the dev numbers
self_dev_number  = self._dev_number  or sys.maxsize
other_dev_number = other._dev_number or sys.maxsize
return self_dev_number > other_dev_number
```

Reply:
Here I checked instead if it's not `None` because the short version gives the wrong result when the `self._dev_numbe` is zero.